### PR TITLE
Implicit GEMM data invokers

### PR DESCRIPTION
**Pull request by DrizztDoUrden**
There are tunable solvers in igemm, so it is a blocker for generic search with invokers (removing run and measure).

I haven't split the backward data invoker to separate. To be honest, I don't feel sure that I understand igemm solvers well enough. And that is not necessary now.

### CODE REVIEW

**Comment by atamazov** [comment]
> blocker for run and measure with invokers.

blocker for "no more RunAndmeasure", actually ;)

**Comment by DrizztDoUrden** [comment]
Updated header)

**Comment by DrizztDoUrden** [comment]
It kind of goes for 1,5 days already :/

**Comment by daniellowell** [comment]
All ready all, reviews please.

File: src/conv/invokers/impl_gemm.cpp [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+                               workspaceDesc,
+                               workSpace,
+                               tensors.outDesc,
+                               tensors.out,
+                               0,
+                               0);
+                    if(handle.IsProfilingEnabled())
+                        elapsed += handle.GetKernelTime();
+                }
+                else if((kernel.GetName() == "gridwise_convolution_implicit_gemm_v4_nchw_kc1x1_"
```
Comment (asroy) [inline src/conv/invokers/impl_gemm.cpp:70]: kernel name is splitted into new line.

Better to keep it on the same line, so it's easy to grep the name

you can use "clang-format-off"

File: src/conv/invokers/impl_gemm.cpp [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+                        if(handle.IsProfilingEnabled())
+                            elapsed += handle.GetKernelTime();
+                    }
+
+                    kernel(tensors.in, tensors.w, tensors.out);
+
+                    if(handle.IsProfilingEnabled())
+                        elapsed += handle.GetKernelTime();
+                }
+                else if(kernel.GetName() == "gridwise_convolution_backward_data_implicit_gemm_v1r1_"
```
Comment (asroy) [inline src/conv/invokers/impl_gemm.cpp:95]: name split

File: src/conv/invokers/impl_gemm.cpp [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+                                            "nchw_kcyx_nkhw" ||
+                        kernel.GetName() == "gridwise_convolution_backward_data_implicit_gemm_v1r1_"
```
Comment (asroy) [inline src/conv/invokers/impl_gemm.cpp:97]: name split

**Comment by DrizztDoUrden** [comment]
Resolved the comments, merged develop.

File: src/conv/invokers/impl_gemm.cpp (hunk 3 above)
Comment (atamazov) [reply src/conv/invokers/impl_gemm.cpp]: @DrizztDoUrden this split is a leftover of clang-format that impedes the programmers from finding kernel names in the code. Please restore strings to undivided state.

@asroy We use ridiculously long names for implicit gemm and I have no idea why. Can we rework filenames and kernel names at once?

Comment (atamazov) [reply src/conv/invokers/impl_gemm.cpp]: @asroy I mean this `gridwise_convolution_backward_data_implicit_gemm` is always the same (except "backward_data" of course), so I would replace it with smth like `gw_conv_fwd/bwd/wrw_igemm` or even `conv_igemm_gw_fwd/bwd/wrw`...

**Comment by atamazov** [comment]
### :cyclone: Testing is in progress...

File: src/conv/invokers/impl_gemm.cpp (hunk 3 above)
Comment (atamazov) [reply src/conv/invokers/impl_gemm.cpp]: > Please restore strings to undivided state.

Oops, already done, missed to reload page.

**Comment by atamazov** [comment]
@DrizztDoUrden Config `./bin/MIOpenDriver conv -x 10 -y 5 -W 341 -H 79 -c 32 -n 4 -k 32 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F 2` has failed with:
```
MIOpen Error: /home/atamazov/github/MLOpen2/src/ocl/convolutionocl.cpp:357: Invoker is not provided by solver ConvHipImplicitGemmBwdDataV4R1
```
Baseline has passed this test. Please find logs in attached [tmp-3.zip](https://github.com/ROCmSoftwarePlatform/MIOpen/files/4540349/tmp-3.zip).

**Comment by DrizztDoUrden** [comment]
I guess I have missed one of the new solvers in the merge process. Should be fixed now.

**Comment by atamazov** [comment]
@asroy thanks for opening #183, this aproach is way to go, I believe ;)

**Comment by atamazov** [comment]
### :cyclone: Testing results of 7c9bbee43 vs 1fecde5f5 baseline

No regressions. Correctness, GPU time, wall time and aux-wall time tested.

- 93 FP32 configs, HIP BE, Normal Find mode and Immediate mode
- 93 FP16 configs, HIP BE, Normal Find mode and Immediate mode
- 93 FP32 configs, OCL BE, Normal Find mode and Immediate mode

Logs available upon request.

**Pull request merged**